stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/repository/domain/message/internal/DatabaseMessageRepository.kt

```diff
@@ -100,18 +105,25 @@ internal class DatabaseMessageRepository(
      */
     override suspend fun insertMessages(messages: List<Message>, cache: Boolean) {
         if (messages.isEmpty()) return
-        val messagesToInsert = messages.flatMap(Companion::allMessages)
-        for (message in messagesToInsert) {
-            require(message.cid.isNotEmpty()) {
-                "message.cid can not be empty. Id of the message: ${message.id}. Text: ${message.text}"
+
+        val validMessages = messages.filter { message -> message.cid.isNotEmpty() }
+
+        // Insert messages.
+        validMessages
+            .onEach { message ->
+                if (messageCache.get(message.id) != null || cache) {
+                    messageCache.put(message.id, message)
+                }
             }
-        }
-        for (m in messagesToInsert) {
-            if (messageCache.get(m.id) != null || cache) {
-                messageCache.put(m.id, m)
+            .map { message -> message.toEntity() }
+            .let { entityMessages -> messageDao.insert(entityMessages) }
+
+        // Insert all replies
+        validMessages.mapNotNull { message -> message.replyTo }
+            .map { message -> message.toReplyEntity() }
+            .let { replyMessages ->
+                replyMessageDao.insert(replyMessages)
             }
```

Comment:
Can a "replyMessage" has another "replyMessage" on it? :thinking: 
Do we need to store it in our db?

Reply:
No, I believe it can't have another reply. 

But that's not the case here. In this snippet all messages that are replies are being saved in `replyMessageDao`, just the messages that are replies.

It is needed to save the replies in a different repository because we can't load all the messages like they are the same thing. If we do so, they will all the on the same list and this will cause the pagination problems that we currently have because the original messages will be loaded without respecting the correct pagination.